agenthub/codeact_agent/codeact_agent.py

```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
         if updated_info:
             for prev_action, obs in updated_info:
                 assert isinstance(
-                    prev_action, (CmdRunAction, AgentEchoAction)
-                ), 'Expecting CmdRunAction or AgentEchoAction for Action'
-                if isinstance(
-                    obs, AgentMessageObservation
-                ):  # warning message from itself
+                    prev_action, self.SUPPORTED_ACTIONS
+                ), f'{prev_action.__class__} is not supported (supported: {self.SUPPORTED_ACTIONS})'
+                # prev_action is already added to self.messages when returned
+
+                # handle observations
+                assert isinstance(
+                    obs, self.SUPPORTED_OBSERVATIONS
+                ), f'{obs.__class__} is not supported (supported: {self.SUPPORTED_OBSERVATIONS})'
+                if isinstance(obs, (AgentMessageObservation, UserMessageObservation)):
                     self.messages.append(
                         {'role': 'user', 'content': obs.content})
+
+                    # User wants to exit
+                    if obs.content.strip() == '/exit':
+                        return AgentFinishAction()
                 elif isinstance(obs, CmdOutputObservation):
                     content = 'OBSERVATION:\n' + obs.content
                     content += f'\n[Command {obs.command_id} finished with exit code {obs.exit_code}]]'
                     self.messages.append({'role': 'user', 'content': content})
+                elif isinstance(obs, IPythonRunCellObservation):
+                    content = 'OBSERVATION:\n' + obs.content
+                    self.messages.append({'role': 'user', 'content': content})
                 else:
                     raise NotImplementedError(
                         f'Unknown observation type: {obs.__class__}'
                     )
+
         response = self.llm.completion(
             messages=self.messages,
-            stop=['</execute>'],
+            stop=[
+                '</execute_ipython>',
+                '</execute_bash>',
+                # chatML in case ollama does not stop
+                '<|im_end|>',
+                '<|im_start|>'
+            ],
             temperature=0.0
         )
         action_str: str = parse_response(response)
-        state.num_of_chars += sum(len(message['content'])
-                                  for message in self.messages) + len(action_str)
+        state.num_of_chars += sum(
+            len(message['content']) for message in self.messages
+        ) + len(action_str)
         self.messages.append({'role': 'assistant', 'content': action_str})
 
-        command = re.search(r'<execute>(.*)</execute>', action_str, re.DOTALL)
-        if command is not None:
+        bash_command = re.search(r'<execute_bash>(.*)</execute_bash>', action_str, re.DOTALL)
+        python_code = re.search(r'<execute_ipython>(.*)</execute_ipython>', action_str, re.DOTALL)
```

Comment:
(nit) we don't have to run regex for `python_code` if `bash_command` matches.

Reply:
done! fixed with walrus operator!